docs/docs/reference/metaprogramming/inline.md

```diff
@@ -0,0 +1,371 @@
+of any of the two dynamic types. The subtype relationship is `B <: A`. Since we
+use the specializing inline syntax, the static types of the vals are inferred
+accordingly. Consequently, calling `meth` on `obj2` is a compile-time error.
+
+```scala
+class A
+class B extends A {
+  def meth() = true
+}
+
+inline def choose(b: Boolean) <: A = {
+  if(b) new A()
+  else new B()
+}
+
+val obj1 = choose(true)  // static type is A
+val obj2 = choose(false) // static type is B
+
+obj1.meth() // compile-time error
+obj2.meth() // OK
+```
+
+```scala
+inline def zero() <: Int = 0
+
+final val one: 1 = zero() + 1
+```
+
+#### Inline Match
+
+A `match` expression in the body of an `inline` method definition may be
+prefixed by the `inline` modifier. If there is enough static information to
+unambiguously take a branch, the expression is reduced to that branch and the
+type of the result is taken. The example below defines an inline method with a
+single inline match expression that picks a case based on its static type:
+
+```scala
+inline def g(x: Any) <: Any = inline x match {
+  case x: String => (x, x) // Tuple2[String, String](x, x)
+  case x: Double => x
+}
+
+g(1.0d): 1.0d
+val t: (String, String) = g("test")
+```
+
+The scrutinee `x` is examined statically and the inline match is reduced
+accordingly returning the corresponding value (with the type specialized due to
+the `<:` in the return type). This example performs a simple type test over the
+scrutinee. The type can have a richer structure like the simple ADT below.
+`toInt` matches the structure of a number in Church-encoding and _computes_ the
+corresponding integer.
+
+```scala
+trait Nat
+case object Zero extends Nat
+case class Succ[N <: Nat](n: N) extends Nat
+
+inline def toInt(n: Nat) <: Int = inline n match {
+  case Zero => 0
+  case Succ(n1) => toInt(n1) + 1
+}
+
+final val two = toInt(Succ(Succ(Zero)))
+val two_ : 2 = two
+```
+
+`two` is inferred to have the singleton type 2.
+
+#### scala.compiletime._
+
+##### Erased Value
+
+We have seen so far inline methods that take terms (tuples and integers) as
+parameters. What if we want to base case distinctions on types instead? For
+instance, one would like to be able to write a function `defaultValue`, that,
+given a type `T` returns optionally the default value of `T`, if it exists. In
+fact, we can already express this using rewrite match expressions and a simple
+helper function, `scala.compiletime.erasedValue`, which is defined as follows:
+
+```scala
+erased def erasedValue[T]: T = ???
+```
+
+The `erasedValue` function _pretends_ to return a value of its type argument
+`T`. In fact, it would always raise a `NotImplementedError` exception when
+called. But the function can in fact never be called, since it is declared
+`erased`, so can be only used a compile-time during type checking.
+
+Using `erasedValue`, we can then define `defaultValue` as follows:
+
+```scala
+inline def defaultValue[T] = inline erasedValue[T] match {
```

Comment:
Put `defaultValue` in another section before `erasedValue`. Describe what it does but do not show the implementation. You can keep this code as an example of how to use `erasedValue` to define `defaultValue`

Reply:
`defaultValue` is merely an example, not part of the `compiletime` package. Does it deserve a section?